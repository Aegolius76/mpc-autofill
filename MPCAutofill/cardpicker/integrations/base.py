from abc import ABC, abstractmethod
from typing import Optional, Type

from cardpicker.models import DFCPair


class InvalidURLException(Exception):
    def __init__(self, import_site_name: str, url: str):
        super().__init__(
            f"There was a problem with importing your list from {import_site_name} at URL {url}. "
            f"Check that your URL is correct and try again."
        )


class ImportSite(ABC):
    """
    Abstract base class for an import site integration. These should facilitate importing a list of cards
    based on a supplied URL.
    """

    @staticmethod
    @abstractmethod
    def get_base_url() -> str:
        """
        Returns the base URL for this import site, e.g. "https://google.com"
        """

        ...

    @classmethod
    @abstractmethod
    def retrieve_card_list(cls, url: str) -> str:
        """
        Takes a URL pointing to a card list hosted on this class's site, queries the site's API / whatever for
        the card list, formats it and returns it.
        """

        ...

    @classmethod
    def raise_invalid_url_exception(cls, url: str) -> None:
        raise InvalidURLException(import_site_name=cls.__name__, url=url)


class GameIntegration(ABC):
    """
    Abstract base class for a game integration. These collect code to integrate with a specific card game
    to enrich the app and tailor it more closely to the community's expectations.
    """

    # region abstract methods

    @classmethod
    @abstractmethod
    def get_dfc_pairs(cls) -> list[DFCPair]:
        ...

    @classmethod
    @abstractmethod
    def get_import_sites(cls) -> list[Type[ImportSite]]:
        ...

    # endregion

    @classmethod
    def query_import_site(cls, url: Optional[str]) -> Optional[str]:
        if url is None:
            raise ValueError("No decklist URL provided.")
        for site in cls.get_import_sites():
            if url.startswith(site.get_base_url()):
                text = site.retrieve_card_list(url)
                cleaned_text = "\n".join(
                    [stripped_line for line in text.split("\n") if len(stripped_line := line.strip()) > 0]
                )
                if len(cleaned_text) > 0:
                    return cleaned_text
        return None


__all__ = ["InvalidURLException", "ImportSite", "GameIntegration"]
