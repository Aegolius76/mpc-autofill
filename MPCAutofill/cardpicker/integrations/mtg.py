import html
import json
import re
from typing import Any, Type

import requests

from cardpicker.integrations.base import GameIntegration, ImportSite
from cardpicker.models import DFCPair
from cardpicker.utils import get_json_endpoint_rate_limited

# region import sites


class Aetherhub(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://aetherhub.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.split("-")[-1]
        response = requests.get(f"{cls.get_base_url()}/Deck/MtgoDeckExport/{deck_id}")
        if response.status_code == 404 or not deck_id:
            cls.raise_invalid_url_exception(url)
        return response.content.decode("utf-8")


class Archidekt(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://archidekt.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        regex_results = re.compile(rf"^{cls.get_base_url()}/decks/(.+?)(?:[\#\/].*)?$").search(url)
        if regex_results is None:
            cls.raise_invalid_url_exception(url)
            return ""  # only necessary so mypy understands the above function throws an exception
        deck_id = regex_results.groups()[0]
        response = requests.get(f"{cls.get_base_url()}/api/decks/{deck_id}/small/")
        if response.status_code == 404 or not deck_id:
            cls.raise_invalid_url_exception(url)
        response_json = json.loads(response.content.decode("utf-8"))
        return "\n".join([f"{x['quantity']} {x['card']['oracleCard']['name']}" for x in response_json["cards"]])


class CubeCobra(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://cubecobra.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        cube_id = url.split("/")[-1]
        response = requests.get(
            f"{cls.get_base_url()}/cube/download/plaintext/{cube_id}"
            f"?primary=Color%20Category"
            f"&secondary=Types-Multicolor&tertiary=Mana%20Value"
            f"&quaternary=Alphabetical"
            f"&showother=false"
        )
        if response.url == "https://cubecobra.com/404" or not cube_id:  # cubecobra returns code 200 for 404 page
            cls.raise_invalid_url_exception(url)
        # filter out lines like `# mainboard` and `# maybeboard` which were recently introduced by cubecobra
        return "\n".join([x for x in response.content.decode("utf-8").split("\n") if not x.startswith("# ")])


class Deckstats(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://deckstats.net"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        regex_results = re.compile(rf"^{cls.get_base_url()}/decks/(\d*)/(\d*)").search(url)
        if regex_results is None:
            cls.raise_invalid_url_exception(url)
            return ""  # only necessary so mypy understands the above function throws an exception
        owner_id, deck_id = regex_results.groups()
        response = requests.get(
            f"{cls.get_base_url()}/api.php"
            f"?action=get_deck"
            f"&id_type=saved"
            f"&owner_id={owner_id}"
            f"&id={deck_id}"
            f"&response_type=list"
        )
        if response.status_code == 404 or not owner_id or not deck_id:
            cls.raise_invalid_url_exception(url)
        card_list = json.loads(response.content.decode("utf-8"))["list"]
        for x in [
            "//Main\n",
            "//Sideboard\n",
            "SB: ",
            "//Instant\n",
            "//Sorcery\n",
            "//Land\n",
            " # !Commander",
        ]:
            card_list = card_list.replace(x, "")
        return card_list


class MagicVille(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://magic-ville.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.strip("#").split("=")[-1]
        response = requests.get(f"{cls.get_base_url()}/fr/decks/dl_appr?ref={deck_id}&save=1")
        card_list = response.content.decode("utf-8")
        for x in ["// www.magic-ville.com deck file\r\n", "SB: "]:
            card_list = card_list.replace(x, "")
        if not deck_id or not card_list:
            cls.raise_invalid_url_exception(url)
        return card_list


class ManaStack(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://manastack.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.split("/")[-1]
        response = requests.get(f"{cls.get_base_url()}/api/deck/list?slug={deck_id}")
        if response.status_code == 404 or not deck_id:
            cls.raise_invalid_url_exception(url)
        response_json = json.loads(response.content.decode("utf-8"))
        card_list = ""
        for x in response_json["list"]["cards"]:
            card_list += f"{x['count']} {x['card']['name']}\n"
        return card_list


class Moxfield(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://www.moxfield.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.split("/")[-1]
        response = requests.get(f"https://api.moxfield.com/v2/decks/all/{deck_id}")

        response_json = json.loads(response.content.decode("utf-8"))
        card_list = ""
        for category in [
            "commanders",
            "companions",
            "mainboard",
            "sideboard",
            "maybeboard",
        ]:
            for name, info in response_json[category].items():
                card_list += f"{info['quantity']} {name}\n"
        for token in response_json["tokens"]:
            if token["layout"] == "token":
                card_list += f"t:{token['name']}\n"
        return card_list


class MTGGoldfish(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://www.mtggoldfish.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.split("#")[0].split("/")[-1]
        response = requests.get(f"{cls.get_base_url()}/deck/download/{deck_id}")
        if response.status_code == 404 or not deck_id:
            cls.raise_invalid_url_exception(url)
        return response.content.decode("utf-8")


class Scryfall(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://scryfall.com"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        deck_id = url.rsplit("#", 1)[0].split("/")[-1]
        response = requests.get(f"https://api.scryfall.com/decks/{deck_id}/export/text")
        if response.status_code == 404 or not deck_id:
            cls.raise_invalid_url_exception(url)
        card_list = response.content.decode("utf-8")
        for x in ["// Sideboard"]:
            card_list = card_list.replace(x, "")
        return card_list


class TappedOut(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://tappedout.net"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        response = requests.get(url + "?fmt=txt")
        if response.status_code == 404:
            cls.raise_invalid_url_exception(url)
        card_list = response.content.decode("utf-8")
        for x in ["Sideboard:\r\n"]:
            card_list = card_list.replace(x, "")
        return card_list


class TCGPlayer(ImportSite):
    @staticmethod
    def get_base_url() -> str:
        return "https://decks.tcgplayer.com/"

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        # TCGPlayer doesn't expose a useful API, so we need to parse the html directly
        response = requests.get(url)
        if response.status_code == 404:
            cls.raise_invalid_url_exception(url)
        card_tuple = re.findall(
            '<span class="subdeck-group__card-qty">(.+?)</span> ' '<span class="subdeck-group__card-name">(.+?)</span>',
            response.text,
        )
        card_list = ""
        for qty, name in card_tuple:
            card_list += "{} {}\n".format(qty, html.unescape(name))
        return card_list


# endregion


class MTG(GameIntegration):
    """
    Our Magic: The Gathering integration reads DFC pairs from Scryfall and enables reading decklists from some
    popular deckbuilding sites.
    """

    DFC_SCRYFALL_QUERY = "is:dfc -layout:art_series -(layout:double_faced_token -keyword:transform) -is:reversible"
    MELD_SCRYFALL_QUERY = "is:meld"
    DFC_SCRYFALL_URL = f"https://api.scryfall.com/cards/search?q={DFC_SCRYFALL_QUERY}"
    MELD_SCRYFALL_URL = f"https://api.scryfall.com/cards/search?q={MELD_SCRYFALL_QUERY}"

    @classmethod
    def query_scryfall_paginated(cls, url: str) -> list[dict[str, Any]]:
        response = get_json_endpoint_rate_limited(url)
        data = response["data"]
        while response["has_more"]:
            response = get_json_endpoint_rate_limited(response["next_page"])
            data += response["data"]
        return data

    @classmethod
    def get_double_faced_card_pairs(cls) -> list[DFCPair]:
        # query data and construct objects for regular double-faced cards
        dfc_pairs: list[DFCPair] = []
        dfc_data = cls.query_scryfall_paginated(cls.DFC_SCRYFALL_URL)
        print(f"Identified {len(dfc_data)} double-faced cards")
        for item in dfc_data:
            if item["digital"] is True:
                continue
            front_name = item["card_faces"][0]["name"]
            back_name = item["card_faces"][1]["name"]
            dfc_pairs.append(DFCPair(front=front_name, back=back_name))
        return dfc_pairs

    @classmethod
    def get_meld_pairs(cls) -> list[DFCPair]:
        # query data and construct objects for meld pairs

        dfc_pairs: list[DFCPair] = []
        meld_data = cls.query_scryfall_paginated(cls.MELD_SCRYFALL_URL)
        print(f"Identified {len(meld_data)} meld pieces")
        for item in meld_data:
            if "all_parts" not in item:
                card_name = item["name"]
                print(f"Skipping {card_name} (missing key 'all_parts')")
                continue

            card_part_singleton_list = list(filter(lambda part: part["name"] == item["name"], item["all_parts"]))
            meld_result_singleton_list = list(
                filter(lambda part: part["component"] == "meld_result", item["all_parts"])
            )

            if len(card_part_singleton_list) != 1 or len(meld_result_singleton_list) != 1:
                continue

            card_part = card_part_singleton_list.pop()
            meld_result = meld_result_singleton_list.pop()["name"]
            if card_part["component"] == "meld_part":
                is_top = "\n(Melds with " not in item["oracle_text"]
                card_bit = "Top" if is_top else "Bottom"
                dfc_pairs.append(DFCPair(front=item["name"], back=f"{meld_result} {card_bit}"))

        return dfc_pairs

    # region implementation of abstract methods

    @classmethod
    def get_dfc_pairs(cls) -> list[DFCPair]:
        return cls.get_double_faced_card_pairs() + cls.get_meld_pairs()

    @classmethod
    def get_import_sites(cls) -> list[Type[ImportSite]]:
        return [
            Aetherhub,
            Archidekt,
            CubeCobra,
            Deckstats,
            MagicVille,
            ManaStack,
            Moxfield,
            MTGGoldfish,
            Scryfall,
            TappedOut,
            TCGPlayer,
        ]

    # endregion


__all__ = ["MTG"]
