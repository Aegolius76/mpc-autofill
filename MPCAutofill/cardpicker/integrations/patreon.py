import operator
import platform
from typing import Optional, TypedDict

import requests

from MPCAutofill.settings import PATREON_ACCESS, PATREON_URL

# Header must be included to access Patreon info
patreon_header = {
    "Authorization": f"Bearer {PATREON_ACCESS}",
    "User-Agent": f"Patreon-Python, version 0.5.1, platform {platform.platform()}",
}


class Campaign(TypedDict):
    # Campaign data scheme
    id: str
    about: str


class Supporter(TypedDict):
    # Patron data scheme
    name: str
    tier: str
    date: str
    usd: int


class SupporterTier(TypedDict):
    # Patron tiers data scheme
    title: str
    description: str
    usd: int


def get_patreon_campaign_details() -> tuple[Optional[Campaign], Optional[dict[str, SupporterTier]]]:
    """
    Get needed patreon campaign details.
    :return: Campaign ID, list of dictionaries containing supporter tier info.
    """

    if not PATREON_URL:
        return None, None

    try:
        res = requests.get(
            # https://docs.patreon.com/#get-api-oauth2-v2-campaigns
            url="https://www.patreon.com/api/oauth2/v2/campaigns",
            params={
                "include": "tiers",
                "fields[campaign]": "summary",
                "fields[tier]": ",".join(["title", "description", "amount_cents"]),
            },
            headers=patreon_header,
        ).json()

        # Properly format campaign details
        campaign: Campaign = {"id": res["data"][0]["id"], "about": res["data"][0]["attributes"]["summary"]}

        # Properly format campaign tiers
        tiers: dict[str, SupporterTier] = {}
        for tier in res["included"]:
            # Ignore free tier
            if tier["attributes"]["amount_cents"] < 1:
                continue
            # Build dictionary of tiers to reference by ID
            tiers[tier["id"]] = {
                "title": tier["attributes"]["title"],
                "description": tier["attributes"]["description"],
                "usd": round(tier["attributes"]["amount_cents"] / 100),
            }
    except KeyError:
        print("Warning: Cannot locate Patreon campaign. Check Patreon access token!")
        return None, None
    return campaign, tiers


def get_patrons(
    campaign_id: str, campaign_tiers: dict[str, SupporterTier], next_page: Optional[str] = None
) -> Optional[list[Supporter]]:
    """
    Get our patreon contributors.
    :return: List of dictionaries containing patreon contributor info.
    """

    if not PATREON_URL:
        return None

    try:
        req = (
            requests.get(
                # https://docs.patreon.com/#get-api-oauth2-v2-campaigns-campaign_id-members
                f"https://www.patreon.com/api/oauth2/v2/campaigns/{campaign_id}/members",
                params={
                    "include": "currently_entitled_tiers",
                    "fields[member]": ",".join(
                        ["full_name", "campaign_lifetime_support_cents", "pledge_relationship_start", "patron_status"]
                    ),
                },
                headers=patreon_header,
            )
            if next_page is None
            else requests.get(next_page, headers=patreon_header)
        )
        res = req.json()

        # Ready the next page if provided
        next_page = res.get("links", {}).get("next")

        # Return formatted list of patrons
        results: list[Supporter] = []
        for mem in res["data"]:

            # Skip non-active members
            if mem["attributes"]["patron_status"] != "active_patron":
                continue

            # Skip members with no tiers
            tiers = mem.get("relationships", {}).get("currently_entitled_tiers", {}).get("data", [])
            if not tiers:
                continue

            # Figure out highest paid tier
            highest: Optional[SupporterTier] = None
            for t in tiers:

                # Skip if tier not recognized
                if t.get("id", "#") not in campaign_tiers:
                    continue

                # This tier is highest
                tier = campaign_tiers[t["id"]]
                if not highest or highest.get("usd", 0) < t["usd"]:
                    highest = SupporterTier(**tier)

            # Skip if no highest tier calculated
            if not highest:
                continue

            # Add member to results
            results.append(
                Supporter(
                    name=mem["attributes"]["full_name"],
                    tier=highest["title"],
                    date=mem["attributes"]["pledge_relationship_start"][:10],
                    usd=highest["usd"],
                )
            )

        # Check if there's additional pages of results
        if next_page:
            results.extend(
                get_patrons(campaign_id=campaign_id, campaign_tiers=campaign_tiers, next_page=next_page) or []
            )

        # Return sorted results
        return sorted(results, key=operator.itemgetter("usd"), reverse=True)

    # Unable to retrieve patrons
    except KeyError:
        print("Warning: Cannot locate Patreon campaign. Check Patreon access token!")
        return None


__all__ = ["Campaign", "Supporter", "SupporterTier", "get_patreon_campaign_details", "get_patrons"]
