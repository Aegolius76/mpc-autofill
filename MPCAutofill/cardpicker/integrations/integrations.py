from typing import Optional, Type

from cardpicker.integrations.base import GameIntegration
from cardpicker.integrations.mtg import MTG

from MPCAutofill.settings import GAME


def get_configured_game_integration() -> Optional[Type[GameIntegration]]:
    integrations: dict[str, Type[GameIntegration]] = {integration.__name__: integration for integration in [MTG]}
    return integrations.get(GAME) if GAME else None
