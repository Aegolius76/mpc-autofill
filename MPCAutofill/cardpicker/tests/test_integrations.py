from collections import Counter
from enum import Enum

import pytest

from cardpicker.integrations.integrations import get_configured_game_integration
from cardpicker.integrations.mtg import MTG


class TestGetIntegration:
    @pytest.mark.parametrize("environment_variable, integration_class", [("MTG", MTG)])
    def test_get_integration(self, db, environment_variable, integration_class, settings):
        settings.GAME = environment_variable
        assert get_configured_game_integration() == integration_class


class TestMTGIntegration:
    # region constants

    class Decks(str, Enum):
        # all of these decks have 4x brainstorm, 3x past in flames, and 1x delver of secrets // insectile aberration
        AETHER_HUB = "https://aetherhub.com/Deck/test-796905"
        ARCHIDEKT = "https://archidekt.com/decks/3380653"
        ARCHIDEKT_WITH_HASH = "https://archidekt.com/decks/3380653#test"
        ARCHIDEKT_WITH_SLASH = "https://archidekt.com/decks/3380653/test"
        CUBE_COBRA = "https://cubecobra.com/cube/overview/2fj4"
        DECK_STATS = "https://deckstats.net/decks/216625/2754468-test"
        MAGIC_VILLE = "https://magic-ville.com/fr/decks/showdeck?ref=948045"
        MANA_STACK = "https://manastack.com/deck/test-426"
        MOXFIELD = "https://www.moxfield.com/decks/D42-or9pCk-uMi4XzRDziQ"
        MTG_GOLDFISH = "https://www.mtggoldfish.com/deck/5149750"
        SCRYFALL = "https://scryfall.com/@mpcautofill/decks/71bb2d40-c922-4a01-a63e-7ba2dde29a5c"
        TAPPED_OUT = "https://tappedout.net/mtg-decks/09-10-22-DoY-test"
        TCG_PLAYER = "https://decks.tcgplayer.com/magic/standard/mpc-autofill/test/1398367"

        def __str__(self):
            return self.value

    # endregion

    # region tests

    def test_get_double_faced_card_pairs(self):
        assert len(MTG.get_double_faced_card_pairs()) > 0

    def test_get_meld_pairs(self):
        assert len(MTG.get_meld_pairs()) > 0

    @pytest.mark.parametrize("url", list(Decks))
    def test_valid_url(self, client, django_settings, snapshot, url):
        decklist = MTG.query_import_site(url)
        assert Counter(decklist.splitlines()) == snapshot

    # endregion
