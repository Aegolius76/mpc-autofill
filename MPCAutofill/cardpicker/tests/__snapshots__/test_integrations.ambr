# name: TestMTGIntegration.test_valid_url[https://aetherhub.com/Deck/test-796905]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://archidekt.com/decks/3380653#test]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://archidekt.com/decks/3380653/test]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://archidekt.com/decks/3380653]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://cubecobra.com/cube/overview/2fj4]
  Counter({
    'Brainstorm': 4,
    'Delver of Secrets': 1,
    'Past in Flames': 3,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://decks.tcgplayer.com/magic/standard/mpc-autofill/test/1398367]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://deckstats.net/decks/216625/2754468-test]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://magic-ville.com/fr/decks/showdeck?ref=948045]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://manastack.com/deck/test-426]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://scryfall.com/@mpcautofill/decks/71bb2d40-c922-4a01-a63e-7ba2dde29a5c]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://tappedout.net/mtg-decks/09-10-22-DoY-test]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://www.moxfield.com/decks/D42-or9pCk-uMi4XzRDziQ]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
    't:Innistrad Checklist': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[https://www.mtggoldfish.com/deck/5149750]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
